zwave_js_server/model/node.py

```diff
@@ -453,17 +453,21 @@ def handle_value_removed(self, event: Event) -> None:
 
     def handle_value_notification(self, event: Event) -> None:
         """Process a node value notification event."""
-        # append metadata if value metadata is available
         value_notification = ValueNotification(self, event.data["args"])
-        value = self.values.get(_get_value_id_from_dict(self, event.data["args"]))
-        if value:
-            value_notification.update(value.data)
-            value.update(event.data["args"])
+        # append metadata if value metadata is available and endpoint
+        # if value notification endpoint isn't provided
+        if value := self.values.get(
+            _get_value_id_from_dict(self, value_notification.data)
+        ):
+            value_notification.metadata.update(value.metadata.data)
+            if value_notification.endpoint is None and value.endpoint is not None:
+                value_notification.data["endpoint"] = value.endpoint
```

Comment:
Would it be safe to first update the ValueNotification with the Value data and then update the ValueNotification with the event data?

We'd avoid accessing details of the data on the ValueNotification that way. It's best if we let each model handle how it should update its data.

Reply:
For reference:
https://github.com/home-assistant-libs/zwave-js-server-python/blob/453318cb752d6be4eb53a243fa595ac5c962da1a/zwave_js_server/model/value.py#L226-L241